Fix incorrect timeReplacement when timescale is large

PhamCongDuc opened the pull request:
We got a problem with floating point arithmetic when calculating time replacement.

Current code does : 
startTime and presentationTimeOffset are devided : 
`timeline.push(
         {start: (startTime / timescale), end: (endTime / timescale)});`
` var pto = (Number(presentationTimeOffset) / timescale) || 0;`
After that they'll multiply to : 
`timeReplacement = (startTime / timescale + info.presentationTimeOffset) * timescale`
If timescale value is large ( such as 10000000), we can get some problem. 
For example:
`unscaleStartTime = 5405266022342269;`
`startTime = unscaleStartTime/10000000; // = 540526602.234227`
`pto = 27994551;`
`presentationTimeOffset = pto/10000000; // = 2.7994551`
`timeReplacement = (startTime + presentationTimeOffset) * 10000000; // = 5405266050336820`
Actually , It should:
`timeReplacement = uncaledStartTime + pto; // = 5405266050336819`
It produces incorrect segment URLs.

To solve that, we should keep unscaled values (unscaledStartTime, unscaledPresentationTimeOffset) when calculating URLs.

PhamCongDuc commented:
Reference to #1 Avoid floating point arithmetic when calculating time replacement

PhamCongDuc commented:
Hi @TobbeEdgeware , @duynguyen-ew , It's OK when I tested it on the latest google/shaka-player (all tests passed and not 404 response from VCP origin).

PhamCongDuc commented:
About testing, I need more time to understand and write test for it. Could I investigate or give up?

ghost commented:
You'll need to describe your changes much better in the commit message. If I were shaka player maintainer I would reject this PR. See https://chris.beams.io/posts/git-commit/ (as basic guideline, not rules)

TobbeEdgeware commented on lib/dash/mpd_utils.js:
```diff
@@ -269,8 +269,13 @@ shaka.dash.MpdUtils.createTimeline = function(
+        end: endTime / timescale,
+        unscaledStart: startTime,
+        unscaledEnd: endTime
+      };
```
Drop the unscaledEnd if it is not needed anywhere.

TobbeEdgeware commented:
@PhamCongDuc 
Regarding commit messages, it is listed as point 5 in https://github.com/google/shaka-player/blob/master/CONTRIBUTING.md with a link to a "guide". You can modify your commit with a more explanatory, but still relatively short message.

I think you have spent enough time on trying to write a test. Let's skip it. I may give it a try before submitting the patch.

My plan is to merge your PR here into our fork, and then submit it tomorrow. lf I don't manage to make a test, I'll simply tell that in the PR towards Shaka.

DucPhamTV committed:
Fix incorrect timeReplacement when large timescale

The current code essentially does this

    timeReplacement =
	(startTime / timescale + presentationTimeOffset) * timescale

When timescale is large enough (e.g. 10 MHz in order to support MS
Smooth Streaming as well), "startTime / timescale * timescale" may not
always be exactly "startTime" because of floating point precision,
which could produce incorrect segment URLs.

Keep startTime and presentationTimeOffset unchanged in the timeline
just to avoid the multiply/divide dance.

PhamCongDuc commented:
Hi @duynguyen-ew , I just changed commit's subject. Please let me know if it's OK. I copied your commit's body which is the best description for it. 
Hi @TobbeEdgeware I removed unscaledEnd.

TobbeEdgeware commented:
I've moved this commit to another branch. Will try to do some more work on it before submitting it to shaka player.

TobbeEdgeware commented:
I submitted a PR today.

lithammer commented:
https://github.com/google/shaka-player/pull/702

TobbeEdgeware commented:
We need to redo this, since the author did @PhamCongDuc is using aa Github account which does not have an edgeware email address.

lithammer commented:
Or just sign the CLA.

TobbeEdgeware commented:
We have signed a CLA for Edgeware, and that is what we should use. We should not have personal contributions and CLAs.
